spec.bs

```diff
@@ -1134,6 +1006,197 @@ To <dfn>convert a modifier to a string</dfn> given a [=part/modifier=] |modifier
+
+  1. Let |dummyURL| be a new [=URL record=].
+  1. Set |dummyURL|'s [=url/scheme=] to |protocolValue|.
+     <p class="note">Note, we set the [=URL record=]'s [=url/scheme=] in order for the [=basic URL parser=] to properly recognize and normalize default port values.</p>
+  1. Let |parseResult| be the result of running [=basic URL parser=] given |portValue| with |dummyURL| as <i>[=basic URL parser/url=]</i> and [=port state=] as <i>[=basic URL parser/state override=]</i>.
+  1. If |parseResult| is failure, then throw a {{TypeError}}.
+  1. Return |dummyURL|'s [=url/port=].
+</div>
+
+<div algorithm>
+  To <dfn>canonicalize a standard pathname</dfn> given a string |value|:
+
+  1. Let |dummyURL| be a new [=URL record=].
+  1. Let |parseResult| be the result of running [=basic URL parser=] given |value| with |dummyURL| as </i>[=basic URL parser/url=]</i> and [=path start state=] as <i>[=basic URL parser/state override=]</i>.
+  1. If |parseResult| is failure, then throw a {{TypeError}}.
+  1. Return |dummyURL|'s [=url/API pathname string=].
+</div>
+
+<div algorithm>
+  To <dfn>canonicalize a cannot-be-a-base-URL pathname</dfn> given a string |value|:
+
+  1. Let |dummyURL| be a new [=URL record=].
+  1. Set |dummyURL|'s [=url/path=][0] to empty string.
+  1. Let |parseResult| be the result of running [=basic URL parser|URL parsing=] given |value| with |dummyURL| as <i>[=basic URL parser/url=]</i> and [=cannot be a base URL path state=] as <i>[=basic URL parser/state override=]</i>.
+  1. If |parseResult| is failure, then throw a {{TypeError}}.
+  1. Return |dummyURL|'s [=url/API pathname string=].
+</div>
+
+<div algorithm>
+  To <dfn>canonicalize a search</dfn> given a string |value|:
+
+  1. Let |dummyURL| be a new [=URL record=].
+  1. Set |dummyURL|'s [=url/query=] to the empty string.
+  1. Let |parseResult| be the result of running [=basic URL parser=] given |value| with |dummyURL| as <i>[=basic URL parser/url=]</i> and [=query state=] as <i>[=basic URL parser/state override=]</i>.
+  1. If |parseResult| is failure, then throw a {{TypeError}}.
+  1. Return |dummyURL|'s [=url/query=].
+</div>
+
+<div algorithm>
+  To <dfn>canonicalize a hash</dfn> given a string |value|:
+
+  1. Let |dummyURL| be a new [=URL record=].
+  1. Set |dummyURL|'s [=url/fragment=] to the empty string.
+  1. Let |parseResult| be the result of running [=basic URL parser=] given |value| with |dummyURL| as <i>[=basic URL parser/url=]</i> and [=fragment state=] as <i>[=basic URL parser/state override=]</i>.
+  1. If |parseResult| is failure, then throw a {{TypeError}}.
+  1. Return |dummyURL|'s [=url/fragment=].
+</div>
+
+<h3 id=canon-processing-for-init>{{URLPatternInit}} Processing</h3>
+
+<div algorithm>
+  To <dfn>process a URLPatternInit</dfn> given a {{URLPatternInit}} |init|, a string |type|, a string |protocol|, a string |username|, a string |password|, a string |hostname|, a string |port|, a string |pathname|, a string |search|, and a string |hash|:
```

Comment:
It sounds like this algorithm is meant to eventually take either strings or null? But it currently is only ever called with strings. If the intent is string-or-null then you should say that instead of string.